File: 1101-1200/1106-Parsing A Boolean Expression/1106-Parsing A Boolean Expression.cpp
class Solution {
public:
    bool parseBoolExpr(string expression) {
        stack<char> St;
        for (char c : expression) {
            if (c == ')') {
                unordered_set<char> visited;
                while (St.top() != '(') {
                    visited.insert(St.top());
                    St.pop();
                }

                St.pop();
                char op = St.top();
                St.pop();
                if (op == '&') {
                    if (visited.count('f')) {
                        St.push('f');
                    }
                    else {
                        St.push('t');
                    }
                }
                else if (op == '|') {
                    if (visited.count('t')) {
                        St.push('t');
                    }
                    else {
                        St.push('f');
                    }
                }
                else if (visited.count('t')) {
                    St.push('f');
                }
                else {
                    St.push('t');
                }
            }
            else if (c != ',') {
                St.push(c);
            }
        }

        return St.top() == 't';
    }
};

File: 1101-1200/1106-Parsing A Boolean Expression/1106-Parsing A Boolean Expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        St = []
        for c in expression:
            if c == ')':
                visited = set()
                while St[-1] != '(':
                    visited.add(St.pop())

                St.pop() # '('
                operator = St.pop()
                if operator == '&':
                    St.append(all(visited))
                elif operator == '|':
                    St.append(any(visited))
                else:
                    St.append(not visited.pop())
            elif c != ',':
                if c == 't':
                    St.append(True)
                elif c == 'f':
                    St.append(False)
                else:
                    St.append(c)

        return St[-1]
        